node/cn/api_tracer.go

```diff
@@ -802,22 +802,31 @@ func (api *PrivateDebugAPI) traceTx(ctx context.Context, message blockchain.Mess
 func (api *PrivateDebugAPI) stateAt(block *types.Block, reexec uint64) (*state.StateDB, func(), error) {
 	var stateDB *state.StateDB
 
-	// If we have the state fully available, use that.
+	// If we have the state fully available in cachedNode, use that.
 	stateDB, err := api.cn.blockchain.StateAtWithGCLock(block.Root())
-	if err != nil {
-		emptyFn := func() {}
+	if err == nil {
+		logger.Debug("Get stateDB from stateCache", "block", block.NumberU64())
+		// During this processing, this lock will prevent to evict the state.
+		return stateDB, stateDB.UnlockGCCachedNode, nil
+	}
 
-		// If no state is locally available, the desired state will be generated.
-		stateDB, err = api.computeStateDB(block, reexec)
-		if err != nil {
-			return nil, emptyFn, err
-		}
-		logger.Debug("Get stateDB by computeStateDB", "block", block.NumberU64())
+	emptyFn := func() {}
+
+	// If we have the state fully available in persistent, use that.
+	stateDB, err = api.cn.blockchain.StateAtWithPersistent(block.Root())
+	if err == nil {
+		logger.Debug("Get stateDB from persistent DB or its cache", "block", block.NumberU64())
 		return stateDB, emptyFn, nil
 	}
-	logger.Debug("Get stateDB from stateCache", "block", block.NumberU64(), "reexec", reexec)
-	// During this processing, this lock will prevent to evict the state.
-	return stateDB, stateDB.UnlockGCCachedNode, nil
+
+	// If no state is locally available, the desired state will be generated.
+	stateDB, err = api.computeStateDB(block, reexec)
```

Comment:
Doesn't it need to compute stateDB with existing state cache?

Reply:
Right. So upper `if` statement deals with that case. :)